### ISSUE TITLE
S2259: False Positive in combination with unrelated ref parameter call

### ISSUE THREAD

**Issue by moellerm** [SonarSource/sonar-dotnet#5285]
### Description

The following example produces a false positive of rule S2259 on the `Length` property access, claiming that the array `a` could be `null`in some cases.

### Repro steps

```
public class Sample {
  public void A(ref string s) {
    s = "a";
  }

  public void B(int i) {
    // empty
  }

  public void C(double[] a) {
    if (a == null) {
      B(0);
    }

    string s = null;
    A(ref s);

    if (a != null) {
      B(a.Length); // S2259: 'a' is null on at least one execution path.
    }
  }
}
```

### Known workarounds

Removing either the if check `if (a == null) { B(0); }` or the call to `A(ref s);` seems to "fix" the problem. But both lines should not have an impact on the code `if (a != null) { B(a.Length); }`.

### Related information

* C#/VB.NET Plugins version: SonarAnalyzer.CSharp 8.33.0.40503 / SonarLint 5.3.0.41207
* Visual Studio version: Visual Studio Pro 2017 (Version 15.9.8)
* Operating System: Windows 10

**Comment by pavel-mikula-sonarsource** [comment]
Hi @moellerm,

Thank you for reporting this interesting case and narrowing it down to a minimal reproducer. I can confirm it as (a tricky) False Positive.
